### ISSUE TITLE
non-deterministic behavior in ontmalizer

### ISSUE THREAD

**Issue by xristy** [srdc/ontmalizer#12]
Build from latest in github and testing with a simple schema and conforming XML instance and using example code from the blog post ([Ontmalizer...](http://www.srdc.com.tr/projects/salus/blog/?p=189)).

I am able to map the schema and then when trying to map instance document, non-deterministically, sometimes a reasonable n3 file is generated and sometimes XML2OWLMapper throws an exception when traversing one of the attributes in the first child element of the instance document.

Repeated trials with the same XMLSchema and sample instance documents sometimes fail and sometimes not, and when there is an exception it is not always at the same attribute - I put Logger calls in XML2OWLMapper on calls to traverseChildren and in traverseAttributes. Here is the exception that is thrown (the stacktrace is below):

> Exception in thread "main" org.apache.jena.ontology.ConversionException: Cannot convert node http://www.srdc.com.tr/ontmalizer#hasValue to OntProperty

The exception, when it occurs, does not occur on the same attribute but always on some attribute of the first child element of the instance document.

The java test, a sample run, schema and instance file are attached as [ontmalizer-nd-issue.zip](https://github.com/srdc/ontmalizer/files/255048/ontmalizer-nd-issue.zip)

**Comment by msfyuksel** [comment]
Hi Chris,

Thanks for reporting this issue. I just wanted to inform you that I am having the same odd behaviour with your instances. So far we have tested ontmalizer with quite complex xml schemas and instances, but never experienced this issue. Yet, I could not identify the real cause of the problem. Hope to find it soon.

Cheers,

**Comment by xristy** [comment]
Hi Mustafa,

I'm glad that the issue is repeatable for you. I'm happy to help try and find the cause. To that end, I should mention that I tried JenaParameters.disableBNodeUIDGeneration(true) in XSD2OWLMapper and in XML2OWLMapper to see if the problem was related to a time-dependent issue related to java.rmi being used to generate UIDs for AnonId/BlankNodeId. Curiously with that setting in XSD2OWLMapper it seemed that XML2OWLMapper always fails. 

I also tried forcing UIDok = false in jena/BlankNodeId.java to force use of jvm local UUIDs while bypassing java.rmi. The non-deterministic behavior seemed to return. 

I'm not sure how to interpret these results but it seems to suggest some sort of blank node id problem. 

Does the problem appear in earlier versions of ontmalizer prior to pull request #11?

I also configured my eclipse test environment to use jena 3.1.1-SNAPSHOT with no change in the issue.

**Comment by msfyuksel** [comment]
Hi Chris,

I finally resolved it. All the time I was looking at wrong places due to non-deterministic behaviour of the issue. The real cause was indeed very simple, as stated in the exception itself: http://www.srdc.com.tr/ontmalizer#hasValue was not explicitly defined as an rdf:Property (or its subclasses) in the ontology created from the xsd. I was able to fix it by just adding 3 letters in the xsd2owl mapper. 

Best,
